### PR TITLE
Adds support for ARGO_CD_[TARGET_REVISION|REVISION] and pass to Custom Tool/Helm/Jsonnet

### PR DESCRIPTION
**Pull request by alexec**
See #740

Changes:

* Fixes a bunch of problems with the docs.
* Adds several doc related make targets (e.g. `make build-docs`).
* Adds support for setting Jsonnet TLAs and Ext Vars from the CLI using `--jsonnet-ext-var-str` and `--jsonnet-tla-str`.
* Adds support for setting Jsonnet TLAs and Ext Vars from the CLI.
* Adds env vars to custom plugins:
  * `ARGOCD_APP_REVISION`
  * `ARGOCD_APP_SOURCE_TARGET_REVISION`
  * `ARGOCD_APP_SOURCE_REPO_URL`
  * `ARGOCD_APP_SOURCE_PATH`
* Allows you to use them for Helm as `--set` and `--set-string` parameters.
* Allows you to use them for Jsonnet as ext vars or TLAs.

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -287,22 +283,29 @@ func GenerateManifests(appPath string, q *apiclient.ManifestRequest) (*apiclient
 	if q.Repo != nil {
 		repoURL = q.Repo.Repo
 	}
+	envVars := v1alpha1.Env{
+		&v1alpha1.EnvEntry{Name:  "ARGOCD_APP_NAME", Value: q.AppLabelValue},
+		&v1alpha1.EnvEntry{Name:  "ARGOCD_APP_NAMESPACE", Value: q.Namespace},
+		&v1alpha1.EnvEntry{Name:  "ARGOCD_APP_TARGET_REVISION", Value: q.ApplicationSource.TargetRevision},
+		&v1alpha1.EnvEntry{Name:  "ARGOCD_APP_REVISION", Value: revision},
+	}
+	postProcessor := envVars.Envsubst()
```
Comment (alexec) [inline reposerver/repository/repository.go]: This is the key change.

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -97,6 +97,16 @@ func (e Env) Environ() []string {
 	return environ
 }
 
+// does an operation similar to `envstubst` tool
+// see https://linux.die.net/man/1/envsubst
+func (e Env) Envsubst(s string) string {
+	for _, e := range e {
+		s = strings.ReplaceAll(s, fmt.Sprintf("$%s", e.Name), e.Value)
+		s = strings.ReplaceAll(s, fmt.Sprintf("${%s}", e.Name), e.Value)
+	}
```
Comment (jannfis) [inline pkg/apis/application/v1alpha1/types.go]: Hm, what about references in `s` to variables that are not set? `envsubst` in Linux silently discards them (i.e. replace them by the empty string `""` , do we want to mimic that behaviour?

**Review by jannfis — CHANGES_REQUESTED** [review]
I only have some minor questions/comments, otherwise looks great! A useful feature to have!

**Comment by alexec** [comment]
Master is currently failing. I'll attempt to understand why that is today.

File: pkg/apis/application/v1alpha1/types.go (hunk 2 above)
Comment (alexec) [reply pkg/apis/application/v1alpha1/types.go]: interesting question, I think this would always be a problem, which you'd need to fix

File: pkg/apis/application/v1alpha1/types.go [hunk 3]
```diff
@@ -270,6 +280,11 @@ type JsonnetVar struct {
 	Code  bool   `json:"code,omitempty" protobuf:"bytes,3,opt,name=code"`
 }
 
+func NewJsonnetVar(s string, code bool) JsonnetVar {
+	parts := strings.SplitN(s, "=", 2)
```
[the remark this answers is not in the document]
Comment (alexec) [reply pkg/apis/application/v1alpha1/types.go]: good point, this currently panics, need to fix

File: ui/src/app/shared/models.ts [hunk 4]
```diff
@@ -175,18 +175,30 @@ export interface ApplicationSourceKsonnet {
     parameters: KsonnetParameter[];
 }
 
-export interface PluginEnv {
+export interface EnvEntry {
```
Comment (alexec) [inline ui/src/app/shared/models.ts]: corrected name to be consistent with `types.go`

File: Makefile [hunk 5]
```diff
@@ -230,3 +230,20 @@ release-precheck: manifests
...
+	mkdocs build
+
+.PHONY: serve-docs
+serve-docs:
+	mkdocs serve
+
+.PHONY: lint-docs
+lint-docs:
+	#  https://github.com/dkhamsing/awesome_bot
+	find docs -name '*.md' -exec grep -l http {} + | xargs docker run --rm -v $(PWD):/mnt:ro dkhamsing/awesome_bot -t 3 --allow-dupe --allow-redirect --white-list `cat white-list | tr "\n" ','` --skip-save-results --
```
Comment (alexec) [inline Makefile]: we had a few broken links - such as ksonnet.io

File: docs/developer-guide/api-docs.md [hunk 6]
```diff
@@ -20,13 +20,12 @@ $ curl $ARGOCD_SERVER/api/v1/applications --cookie "argocd.token=$ARGOCD_TOKEN"
...
+> v1.3
 
 Then pass using the HTTP `Authorization` header, prefixing with `Bearer `:
 
 ```bash
 $ curl $ARGOCD_SERVER/api/v1/applications -H "Authorization: Bearer $ARGOCD_TOKEN" 
 {"metadata":{"selfLink":"/apis/argoproj.io/v1alpha1/namespaces/argocd/applications","resourceVersion":"37755"},"items":...}
 ```
- 
- You sh
```
Comment (alexec) [inline docs/developer-guide/api-docs.md]: ops!

File: pkg/apis/application/v1alpha1/types.go (hunk 2 above)
Comment (alexec) [reply pkg/apis/application/v1alpha1/types.go]: I've decided not to change this. Fix is complex.

**Review by jannfis — APPROVED** [review]

File: ui/src/app/shared/components/env-input/env-input.tsx [hunk 7]
```diff
@@ -36,7 +36,7 @@ class State {
     public newItem: Item;
 }
 
-class ArrayInput extends React.Component<Props, State> {
+class EnvInput extends React.Component<Props, State> {
     constructor(props: Readonly<Props>) {
         super(props);
         this.state = {newItem: {name: '', value: ''}, items: props.items};
```
Comment (alexmt) [inline ui/src/app/shared/components/env-input/env-input.tsx]: I missed that when ArrayInput was introduced. This component does not require state since state is stored in the Form. State should be removed and items from properties should be used to render the component.

File: ui/src/app/shared/components/vars-input/vars-input.tsx [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import * as React from 'react';
+import * as ReactForm from 'react-form';
+
+/*
+This is a lift-and-shift of env-input.
```
Comment (alexmt) [inline ui/src/app/shared/components/vars-input/vars-input.tsx]: Please reuse ArrayInput. You might use generic to support items type and support custom input renderer. e.g.:

```
export const ArrayInput<T> = (props: {items: []T, renderInput: (item: T): ReactNode ) {

```

File: white-list [hunk 9]
```diff
@@ -0,0 +1,22 @@
+10.97.164.88
```
Comment (alexmt) [inline white-list]: why is this file required?

File: ui/src/app/shared/components/env-input/env-input.tsx [hunk 10]
```diff
@@ -36,7 +36,7 @@ class State {
     public newItem: Item;
 }
 
-class ArrayInput extends React.Component<Props, State> {
```
Comment (alexmt) [inline ui/src/app/shared/components/env-input/env-input.tsx]: ArrayInput is already used to edit application labels

File: docs/user-guide/build-environment.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# Build Environment
+
+[Custom tools](config-management-plugins.md), [Helm](helm.md), and [Jsonnet](jsonnet.md) support the following build env vars:
+
+* `ARGOCD_APP_NAME` - name of application
+* `ARGOCD_APP_NAMESPACE` - destination application namespace.
+* `ARGOCD_APP_TARGET_REVISION` - the target revision from the spec, e.g. `master`.
+* `ARGOCD_APP_REVISION` - the resolved revision, e.g. `f913b6cbf58aa5ae5ca1f8a2b149477aebcbd9d8`
```
Comment (alexmt) [inline docs/user-guide/build-environment.md]: Why only config managed plugins have access to system environment variables but not Helm and Jsonnet? I think we should be consistent and provide system environment variables to all tools.

**Review by alexmt — DISMISSED** [review]

File: docs/user-guide/build-environment.md (hunk 11 above)
Comment (alexec) [reply docs/user-guide/build-environment.md]: I don't believe they can consume them?

File: ui/src/app/shared/components/vars-input/vars-input.tsx (hunk 8 above)
Comment (alexec) [reply ui/src/app/shared/components/vars-input/vars-input.tsx]: I was scared by genercising it , I thought i would make them too  complex. Lets discus IRL.

File: docs/user-guide/build-environment.md (hunk 11 above)
Comment (alexmt) [reply docs/user-guide/build-environment.md]: Why not? In case of helm user might want to use 

```yaml
  spec:
    source:
      helm:
        parameters:
        - name: my-parameter
          value: $MY_ENV_VARIABLE
```

File: white-list (hunk 9 above)
Comment (alexec) [reply white-list]: a list of site we ignore when checking for broken links in mkdocs

File: docs/user-guide/build-environment.md (hunk 11 above)
Comment (alexec) [reply docs/user-guide/build-environment.md]: ah - that is how they work - do you want a demo?

**Comment by alexec** [comment]
@alexmt this is ready for review please

File: test/e2e/jsonnet_test.go [hunk 12]
```diff
@@ -72,3 +73,36 @@ func TestJsonnetTlaParameterAppliedCorrectly(t *testing.T) {
 			assert.Equal(t, int64(3), *kube.GetDeploymentReplicas(deployment))
 		})
 }
+
+func TestJsonnetTlaEnv(t *testing.T) {
+	Given(t).
+		Path("jsonnet-tla-cm").
+		JsonnetTLAStrParameter("foo=$ARGOCD_APP_NAME").
```
Comment (alexec) [inline test/e2e/jsonnet_test.go]: should probably change into parameter to `Create()`

**Comment by alexec** [comment]
@alexmt this is ready for review, but could be broken up into 4x small PRs if you like.

File: ui/src/app/shared/components/array-input/array-input-field.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
+import * as React from 'react';
```
Comment (alexmt) [inline ui/src/app/shared/components/array-input/array-input-field.tsx]: I've sent PR which refactors array input: removes duplicated state and reduce duplication: https://github.com/alexec/argo-cd/pull/10 . Let me know what you think please.

**Review by alexmt — DISMISSED** [review]

**Review by alexmt — APPROVED** [review]
